File: app/src/main/java/com/meteoro/testeandroid/core/view/BaseActivity.java
package com.meteoro.testeandroid.core.view;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

import com.meteoro.testeandroid.core.di.module.ActivityModule;

import icepick.Icepick;

public abstract class BaseActivity extends AppCompatActivity {

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Icepick.restoreInstanceState(this, savedInstanceState);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        Icepick.saveInstanceState(this, outState);
    }

    public ActivityModule getActivityModule() {
        return new ActivityModule(this);
    }
}

File: app/src/main/java/com/galdino/testandroid/screens/base/BaseActivity.kt
package com.galdino.testandroid.screens

import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem
import android.widget.Toast
import com.galdino.testandroid.mvp.Contract
import com.galdino.testandroid.screens.base.ToastController

abstract class BaseActivity: AppCompatActivity(), Contract.View {
    private lateinit var mToastController: ToastController

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(getLayoutResource())
        mToastController = getToastController()
        onInitView()
    }

    open fun getToastController() = ToastController(this)

    abstract fun getLayoutResource(): Int

    abstract fun onInitView()

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        if (item?.itemId == android.R.id.home) {
            onBackPressed()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    fun showShortToast(resId: Int) {
        mToastController.showShortToast(resId)
    }

    fun showShortToast(text: String?) {
        mToastController.showToast(text, Toast.LENGTH_SHORT)
    }

    fun showLongToast(resId: Int) {
        mToastController.showShortToast(resId)
    }

    fun showLongToast(text: String?) {
        mToastController.showToast(text, Toast.LENGTH_LONG)
    }

    override fun onDestroy() {
        mToastController.cancel()
        super.onDestroy()
    }
}